### PR TITLE
Add vast event trackers support for EventType

### PR DESCRIPTION
**Pull request by marki1an**
https://github.com/prebid/prebid-server/pull/2179

### CODE REVIEW

File: src/main/java/org/prebid/server/events/EventUtil.java [hunk 1]
```diff
@@ -39,15 +42,33 @@
 
     private static final String LINE_ITEM_ID_PARAMETER = "l";
 
+    private static final List<String> POSSIBLE_TYPES = List.of(WIN_TYPE, IMP_TYPE, VAST_TYPE);
+    private static final List<String> POSSIBLE_V_TYPES = List.of(EventRequest.VastType.START.toString(),
```
Comment (And1sS) [inline src/main/java/org/prebid/server/events/EventUtil.java:46]: These two must be sets.

File: src/main/java/org/prebid/server/events/EventRequest.java [hunk 2]
```diff
@@ -30,9 +30,11 @@
 
     String lineItemId;
 
+    VastType vType;
+
     public enum Type {
 
-        win, imp
+        win, imp, VAST
```
Comment (And1sS) [inline src/main/java/org/prebid/server/events/EventRequest.java:37]: Rename all to UPPERCASE, or rename VAST to lowercase. Names of values in enum should have similar styling.

File: src/main/java/org/prebid/server/events/EventRequest.java [hunk 3]
```diff
@@ -44,4 +46,28 @@
...
+        FIRST_QUARTILE("firstQuartile"),
+        MID_POINT("midPoint"),
+        THIRD_QUARTILE("thirdQuartile"),
+        COMPLETE("complete");
+
+        private final String name;
+
+        VastType() {
+            this.name = name();
+        }
```
Comment (And1sS) [inline src/main/java/org/prebid/server/events/EventRequest.java:62]: This one is redundant

File: src/main/java/org/prebid/server/events/EventUtil.java [hunk 4]
```diff
@@ -39,15 +42,33 @@
...
+            EventRequest.VastType.FIRST_QUARTILE.toString(),
+            EventRequest.VastType.MID_POINT.toString(),
+            EventRequest.VastType.THIRD_QUARTILE.toString(),
+            EventRequest.VastType.COMPLETE.toString());
```
Comment (And1sS) [inline src/main/java/org/prebid/server/events/EventUtil.java:50]: What about making this utility function(maybe extract this to EnumUtil?)?:
```
private static <E extends Enum<E>> Set<String> stringValues(Class<E> enumeration) {
        return EnumSet.allOf(enumeration).stream()
                .map(Enum::toString)
                .collect(Collectors.toSet());
    }
```

then we can simplify:
```
private static final Set<String> POSSIBLE_TYPES = stringValues(EventRequest.Type.class);
private static final Set<String> POSSIBLE_V_TYPES = stringValues(EventRequest.VastType.class);
```

File: src/main/java/org/prebid/server/events/EventUtil.java [hunk 5]
```diff
@@ -39,15 +42,33 @@
...
+
     private EventUtil() {
     }
 
     public static void validateType(RoutingContext routingContext) {
         final String type = routingContext.request().params().get(TYPE_PARAMETER);
-        if (ObjectUtils.notEqual(type, IMP_TYPE) && ObjectUtils.notEqual(type, WIN_TYPE)) {
+        final boolean isNotEqual = POSSIBLE_TYPES.stream().allMatch(element -> ObjectUtils.notEqual(element, type));
+        if (isNotEqual) {
```
Comment (And1sS) [inline src/main/java/org/prebid/server/events/EventUtil.java:58]: Can be simplified to:
```
final String type = StringUtils.defaultString(routingContext.request().params().get(TYPE_PARAMETER));
if (POSSIBLE_TYPES.contains(type)) {
```

File: src/main/java/org/prebid/server/events/EventUtil.java [hunk 6]
```diff
@@ -39,15 +42,33 @@
...
+            throw new IllegalArgumentException(String.format(
+                    "Type '%s' is required query parameter. Possible values are %s, but was %s",
+                    TYPE_PARAMETER, POSSIBLE_TYPES, type));
+        }
+    }
+
+    public static void validateVType(RoutingContext routingContext) {
+        final String vType = routingContext.request().params().get(V_TYPE_PARAMETER);
+        final boolean isNotEqual = POSSIBLE_V_TYPES.stream().allMatch(element -> ObjectUtils.notEqual(element, vType));
+        if (isNotEqual) {
```
Comment (And1sS) [inline src/main/java/org/prebid/server/events/EventUtil.java:68]: Same here

**Review by And1sS — CHANGES_REQUESTED** [review]

**Review by And1sS — DISMISSED** [review]

File: src/main/java/org/prebid/server/events/EventUtil.java [hunk 7]
```diff
@@ -39,15 +43,27 @@ public class EventUtil {
...
+                    "Type '%s' is required query parameter. Possible values are %s, but was %s",
+                    TYPE_PARAMETER, POSSIBLE_TYPES, type));
+        }
+    }
+
+    public static void validateVType(RoutingContext routingContext) {
+        final String vType = routingContext.request().params().get(V_TYPE_PARAMETER);
+        if (!POSSIBLE_V_TYPES.contains(vType)) {
+            throw new IllegalArgumentException(String.format(
+                    "Type '%s' is required query parameter. Possible values are %s, but was %s",
```
Comment (CTMBNara) [inline src/main/java/org/prebid/server/events/EventUtil.java:65]: `vtype` parameter is not always required. As you can see [here](https://github.com/prebid/prebid-server/pull/2179/files#diff-ed708c8cdfaf7ce0ada46a2901228d5a2ed5d6d9bc59981b3eda6ee02111bd19R148) it is required iff `eventType == VAST`.

**Review by CTMBNara — CHANGES_REQUESTED** [review]

File: src/main/java/org/prebid/server/events/EventUtil.java (hunk 7 above)
Comment (marki1an) [reply src/main/java/org/prebid/server/events/EventUtil.java:65]: Yeah... you're right, sorry missed this case

File: src/main/java/org/prebid/server/events/EventUtil.java [hunk 8]
```diff
@@ -39,15 +43,34 @@ public class EventUtil {
...
+        final String vType = routingContext.request().params().get(V_TYPE_PARAMETER);
+
+        if (type.equals(VAST_TYPE)) {
+            if (!POSSIBLE_V_TYPES.contains(vType)) {
+                throw new IllegalArgumentException(String.format(
+                        "Type '%s' is required query parameter. Possible values are %s, but was %s",
+                        V_TYPE_PARAMETER, POSSIBLE_V_TYPES, vType));
+            }
+        } else if (StringUtils.isNotBlank(vType)) {
+            throw new IllegalArgumentException(String.format("Type '%s' is only required for when '%s=%s'",
```
Comment (CTMBNara) [inline src/main/java/org/prebid/server/events/EventUtil.java:72]: `String.format("Parameter '%s' is only required for %s=%s", V_TYPE_PARAMETER, TYPE_PARAMETER, VAST_TYPE)`

File: src/test/java/org/prebid/server/handler/NotificationEventHandlerTest.java [hunk 9]
```diff
@@ -124,7 +124,8 @@ public void shouldReturnBadRequestWhenBidIdIsMissing() {
         // given
         given(httpRequest.params())
                 .willReturn(MultiMap.caseInsensitiveMultiMap()
-                        .add("t", "win"));
+                        .add("t", "win")
+                        .add("vtype", "start"));
```
Comment (CTMBNara) [inline src/test/java/org/prebid/server/handler/NotificationEventHandlerTest.java:128]: Redundant

File: src/test/java/org/prebid/server/handler/NotificationEventHandlerTest.java [hunk 10]
```diff
@@ -140,6 +141,7 @@ public void shouldReturnBadRequestWhenTimestampIsInvalid() {
         // given
         given(httpRequest.params()).willReturn(MultiMap.caseInsensitiveMultiMap()
                 .add("t", "win")
+                .add("vtype", "start")
```
Comment (CTMBNara) [inline src/test/java/org/prebid/server/handler/NotificationEventHandlerTest.java:144]: Redundant

File: src/test/java/org/prebid/server/handler/NotificationEventHandlerTest.java [hunk 11]
```diff
@@ -175,6 +177,7 @@ public void shouldReturnBadRequestWhenFormatValueIsInvalid() {
...
```
Comment (CTMBNara) [inline src/test/java/org/prebid/server/handler/NotificationEventHandlerTest.java:180]: Redundant

File: src/test/java/org/prebid/server/handler/NotificationEventHandlerTest.java [hunk 12]
```diff
@@ -193,6 +196,7 @@ public void shouldReturnBadRequestWhenAnalyticsValueIsInvalid() {
...
```
Comment (CTMBNara) [inline src/test/java/org/prebid/server/handler/NotificationEventHandlerTest.java:199]: Redundant

File: src/test/java/org/prebid/server/handler/NotificationEventHandlerTest.java [hunk 13]
```diff
@@ -212,6 +216,7 @@ public void shouldReturnBadRequestWhenIntegrationValueIsInvalid() {
...
```
Comment (CTMBNara) [inline src/test/java/org/prebid/server/handler/NotificationEventHandlerTest.java:219]: Redundant

File: src/test/java/org/prebid/server/handler/NotificationEventHandlerTest.java [hunk 14]
```diff
@@ -426,6 +431,7 @@ public void shouldNotPassEventToAnalyticsReporterWhenAnalyticsValueIsZero() {
...
```
Comment (CTMBNara) [inline src/test/java/org/prebid/server/handler/NotificationEventHandlerTest.java:434]: Redundant

**Review by CTMBNara — CHANGES_REQUESTED** [review]

File: src/main/java/org/prebid/server/events/EventUtil.java [hunk 15]
```diff
@@ -39,15 +43,34 @@ public class EventUtil {
...
+                    TYPE_PARAMETER, POSSIBLE_TYPES, type));
+        }
+    }
+
+    public static void validateVType(RoutingContext routingContext) {
+        final String type = routingContext.request().params().get(TYPE_PARAMETER);
+        final String vType = routingContext.request().params().get(V_TYPE_PARAMETER);
+
+        if (type.equals(VAST_TYPE)) {
+            if (!POSSIBLE_V_TYPES.contains(vType)) {
```
Comment (And1sS) [inline src/main/java/org/prebid/server/events/EventUtil.java:66]: Why nesting two `if`'s? Can be a single logical expression

**Review by And1sS — APPROVED** [review]

**Comment by SerhiiNahornyi** [comment]
Needs video functionality before creation

**Pull request closed without merging**